### DISCUSSION TITLE
Vitest testing Error: atom.read, atom.write is not a function

### DISCUSSION THREAD

**Discussion by n8urm8** [pmndrs/jotai#2643] [Q&A] [answered]
Hello, I'm having some issues running vitest testing on some components that use an atom that I updated to a different pattern, but was properly passing tests previously. Now, I get this error `TypeError: atom.read is not a function`.  I don't need to hydrate any atoms for the component to work, and I'll I'm trying to test is that the component renders. If I do hydrate the atoms with mock data, I get a similar error `TypeError: atom.write is not a function`.
Here's the code of the atom.
```
export const filterValueStateAtom = atom<Record<string, readonly FiltersType[]>>({});

export const selectedFilterValueState = (rateSetId: string, tabId: DrilldownType) => {
  const key = `${rateSetId}-${tabId}`; 

  const selectedFilterValuesAtom = atom(
    (get) => get(filterValueStateAtom)[key] ?? [{ id: '', name: '', relatedKeys: [] }],
    (get, set, value: readonly FiltersType[]) => {
      const storage = get(filterValueStateAtom);
      const newStorage = { ...storage, [key]: value };
      set(filterValueStateAtom, newStorage);
    }
  );
  return selectedFilterValuesAtom;
};
```

Here's the implementation:
` const filterValues = useAtomValue(
    useMemo(() => selectedFilterValueState(rateSetId ?? '', tabId ?? 'historical-rates'), [rateSetId, tabId])
  );`
  
  I have been using this pattern in other atoms for a while, and it's been working fine. The change that caused this to return an error in testing is that previously I just used a hook for `selectedFilterValuesAtom` to do the exact same thing, just not as an atom. The app runs without errors, just getting an error in tests. Running vite, react 19. If anyone has any insight, I'd be greatly appreciative.

**Comment by n8urm8** [comment, marked as answer]
Found my issue, I had created a mock implementation incorrectly, and now it's all good.

**Reply by BlakeEriks** [reply]
Im getting this issue currently. Do you know specifically what you changed?

**Reply by Lysander086** [reply]
Yes, appreciate if @n8urm8 could provide the correct mocking way.
